src/data/sparse_page_dmatrix.cc

```diff
@@ -12,261 +12,91 @@
-    std::vector<std::unique_ptr<dmlc::SeekStream> >&& files)
-    : page_(nullptr), clock_ptr_(0), files_(std::move(files)) {
-  load_all_ = false;
-  formats_.resize(files_.size());
-  prefetchers_.resize(files_.size());
-
-  for (size_t i = 0; i < files_.size(); ++i) {
-    dmlc::SeekStream* fi = files_[i].get();
-    std::string format;
-    CHECK(fi->Read(&format)) << "Invalid page format";
-    formats_[i].reset(SparsePageFormat::Create(format));
-    SparsePageFormat* fmt = formats_[i].get();
-    size_t fbegin = fi->Tell();
-    prefetchers_[i].reset(new dmlc::ThreadedIter<SparsePage>(4));
-    prefetchers_[i]->Init([this, fi, fmt] (SparsePage** dptr) {
-        if (*dptr == nullptr) {
-          *dptr = new SparsePage();
-        }
-        if (load_all_) {
-          return fmt->Read(*dptr, fi);
-        } else {
-          return fmt->Read(*dptr, fi, index_set_);
-        }
-      },  [this, fi, fbegin] () {
-        fi->Seek(fbegin);
-        index_set_ = set_index_set_;
-        load_all_ = set_load_all_;
-      });
-  }
-}
-
-SparsePageDMatrix::ColPageIter::~ColPageIter() {
-  delete page_;
+MetaInfo& SparsePageDMatrix::Info() {
+  return row_source_->info;
 }
 
-bool SparsePageDMatrix::ColPageIter::Next() {
-  // doing clock rotation over shards.
-  if (page_ != nullptr) {
-    size_t n = prefetchers_.size();
-    prefetchers_[(clock_ptr_ + n - 1) % n]->Recycle(&page_);
-  }
-  if (prefetchers_[clock_ptr_]->Next(&page_)) {
-    // advance clock
-    clock_ptr_ = (clock_ptr_ + 1) % prefetchers_.size();
-    return true;
-  } else {
-    return false;
-  }
+const MetaInfo& SparsePageDMatrix::Info() const {
+  return row_source_->info;
 }
 
-void SparsePageDMatrix::ColPageIter::BeforeFirst() {
-  clock_ptr_ = 0;
-  for (auto& p : prefetchers_) {
-    p->BeforeFirst();
+class SparseBatchIteratorImpl : public BatchIteratorImpl {
+ public:
+  explicit SparseBatchIteratorImpl(SparsePageSource* source)
+      : source_(source) {}
+  const SparsePage& operator*() const override { return source_->Value(); }
+  void operator++() override { at_end_ = !source_->Next(); }
+  bool AtEnd() const override { return at_end_; }
+  SparseBatchIteratorImpl* Clone() override {
+    return new SparseBatchIteratorImpl(*this);
   }
-}
-
-void SparsePageDMatrix::ColPageIter::Init(
-    const std::vector<bst_uint>& index_set) {
-  set_index_set_ = index_set;
-  set_load_all_ = true;
-  std::sort(set_index_set_.begin(), set_index_set_.end());
-  this->BeforeFirst();
-}
 
-  dmlc::DataIter<SparsePage>* SparsePageDMatrix::ColIterator() {
-  CHECK(col_iter_ != nullptr);
-  std::vector<bst_uint> col_index;
-  std::iota(col_index.begin(), col_index.end(), bst_uint(0));
-  col_iter_->Init(col_index);
-  return col_iter_.get();
+ private:
+  SparsePageSource* source_{nullptr};
+  bool at_end_{ false };
+};
+
+BatchSet SparsePageDMatrix::GetRowBatches() {
+  auto cast = dynamic_cast<SparsePageSource*>(row_source_.get());
+  cast->BeforeFirst();
+  cast->Next();
```

Comment:
I am using the DataSource class to do the actual iteration, these two lines ensure it is reset to the first row batch, which otherwise may not be true.